src/shared/bus-unit-util.c

```diff
@@ -431,7 +431,7 @@ static int bus_append_cgroup_property(sd_bus_message *m, const char *field, cons
                          * ones). This way the physical memory size can be determined server-side */
 
                         n = strjoina(field, "Scale");
-                        r = sd_bus_message_append(m, "(sv)", n, "u", (uint32_t) (((uint64_t) UINT32_MAX * r) / 100U));
+                        r = sd_bus_message_append(m, "(sv)", n, "u", (uint32_t) (((uint64_t) r * (uint64_t) UINT32_MAX) / UINT64_C(1000)));
```

Comment:
All that casting is not necessary. The shorter operand is always up-cast to the longer type automatically, and I don't think there's any reason to duplicate what the compiler will do anyway.

Reply:
the casting is necessary to some point. we want to do the multiplication/division in 64bit, and we want to pass 32bit value (and as this is varargs we need to be careful to put the right type on the stack). 

But you are right to some point, this would suffice too, I guess:
```c
(uint32_t) (((uint64_t) r * UINT32_MAX) / 1000)
```

There is one reason I wrote this the way I did though: i find the C signed/unsigned propagation rules quite counter-intuitive as it's the "unsigned" that wins, not the "signed" if you multiply/divide signed with unsigned. In the syntax I chose this C behaviour is irrelevant, as we never multiply/divide two values of different signedness with each other.

Reply:
So I think
> `(uint32_t) (((uint64_t) r * UINT32_MAX) / 1000)`

is nice… That's more-or-less what I was expecting.

(I'm complaining about the casts because it's just too easy to add casts everywhere, making the interesting bits harder to read.)